pkg/tarutil/tarutil.go

```diff
@@ -55,3 +69,118 @@ func FindTarBinary() (string, bool, error) {
 	}
 	return "", false, fmt.Errorf("failed to find `tar` binary")
 }
+
+type Tarballer struct {
+	Buffer    *bufio.Writer
+	TarWriter *tar.Writer
+	seenFiles map[uint64]string
+}
+
+// TODO: Add tar options for compression, whiteout files, chown ..etc
+
+func NewTarballer(writer io.Writer) *Tarballer {
```

Comment:
Probably you can just use https://pkg.go.dev/github.com/containerd/containerd/archive#WriteDiff .
(input `A` = container rootfs, input `B` = an empty dir)